### PR TITLE
buddyinfo plugin: new plugin for memory fragmentation info

### PR DESCRIPTION
**Pull request by asafkahlon**
ChangeLog: buddyinfo plugin: New plugin for memory fragmentation info.

Signed-off-by: Asaf Kahlon <asafka7@gmail.com>

### CODE REVIEW

**Comment by mrunge** [comment]
Thank you for your contribution to collectd!

Before going through this in more detail, you have copyrights in there from people not being you. Is that correct?

**Comment by asafkahlon** [comment]
> Before going through this in more detail, you have copyrights in there from people not being you. Is that correct?

Oops :)
I'm sorry, I probably missed that when I did a naive copy-paste from one of the other plugins. I'll push a correction for that.

File: src/buddyinfo.c [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+#define NUM_OF_KB(pagesize, order) ((pagesize) / 1024) * (1 << (order))
+
+static const char *config_keys[] = {"Zone"};
+static int config_keys_num = STATIC_ARRAY_SIZE(config_keys);
+
+static ignorelist_t *ignorelist;
+
+static int buddyinfo_config(const char *key, const char *value) {
+  if (ignorelist == NULL)
+    ignorelist = ignorelist_create(1);
```
Comment (kwiatrox) [inline src/buddyinfo.c]: Add check after ignorelist_create if ignorelist != NULL

File: src/buddyinfo.c [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+  if (strcasecmp(key, "Zone") == 0)
+    ignorelist_add(ignorelist, value);
```
Comment (kwiatrox) [inline src/buddyinfo.c]: Check if returned 0 (on success)

File: src/buddyinfo.c [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  else
+    return -1;
```
Comment (kwiatrox) [inline src/buddyinfo.c]: Add ERROR about wrong option.

File: src/buddyinfo.c [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+    numfields = strsplit(dummy, fields, BUDDYINFO_FIELDS);
+    if (numfields != BUDDYINFO_FIELDS)
+      continue;
+
+    zone = fields[1];
+    for (int i = 1; i <= MAX_ORDER; i++) {
+      ssnprintf(pagesize_kb, sizeof(pagesize_kb), "%dKB",
+                NUM_OF_KB(pagesize, i - 1));
+      buddyinfo_submit(zone, pagesize_kb, atoll(fields[i + 1]));
```
Comment (kwiatrox) [inline src/buddyinfo.c]: cast from long long to int

File: src/buddyinfo.c [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  fclose(fh);
+  return 0;
+}
+
+void module_register(void) {
+  plugin_register_config("buddyinfo", buddyinfo_config, config_keys,
+                         config_keys_num);
+  plugin_register_read("buddyinfo", buddyinfo_read);
```
Comment (kwiatrox) [inline src/buddyinfo.c]: You need also plugin_register_shutdown to clean ignorelist with `ignorelist_free`.

File: src/collectd.conf.pod [hunk 6]
```diff
@@ -1455,6 +1455,20 @@ By default no detailed zone information is collected.
...
+
+The B<buddyinfo> plugin collects information by reading "/proc/buddyinfo".
+This file contains information about the number of available contagious
+physical pages at the moment.
+
+=over 4
+
+=item B<Zone> I<ZoneName>
+
+Zone to collect info about. Will collect all zones by default.
```
Comment (kwiatrox) [inline src/collectd.conf.pod]: Optional: leaving the string empty like `Zone ""` should also result in default behavior.

File: src/buddyinfo.c [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    return -1;
+  }
+
+  while (fgets(buffer, sizeof(buffer), fh) != NULL) {
+    if (!(dummy = strstr(buffer, "zone")))
+      continue;
+
+    numfields = strsplit(dummy, fields, BUDDYINFO_FIELDS);
+    if (numfields != BUDDYINFO_FIELDS)
+      continue;
```
Comment (kwiatrox) [inline src/buddyinfo.c]: Is it normal and expected to have less fields?, if not add some WARNING/INFO here for easier debugging of problems.

**Comment by asafkahlon** [comment]
> Thank you for your contribution.
> 
> 1. Why do you omit Node info? There can be more than one and depending on the use case it may be beneficial to add this info to plugin-instance. My example below:
> 
> ```
> Node 0, zone   Normal   1618   2566   3331   3607   1383    776    756    295     44     33  30127
> Node 1, zone   Normal    934   2174   2477   1909   1377    861    400    268    144     78  30827
> ```
> 
You're right. The current implementation doesn't contain node filtering.
I can add it to this PR if it's really needed, but I thought zone filtering is more important as a first stage and nodes can be added on a different PR (again, can be added here if needed).
BTW, I already used the zone name for plugin-instance so I'm not sure where this field should go...

> 2. What is the use case? You might consider use of /proc/pagetypeinfo instead of buddyinfo to get more detailed information.

The use case is giving the ability to track and debug the fragmentation status on the system.
In my opinion pagetypeinfo contains many information that maybe won't interest the average user (maybe it worth a separate plugin...?)

Anyway, thanks for the review!
I fixed your comments about the code.

File: src/buddyinfo.c [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+static const char *config_keys[] = {"Zone"};
+static int config_keys_num = STATIC_ARRAY_SIZE(config_keys);
+
+static ignorelist_t *ignorelist;
+
+static int buddyinfo_config(const char *key, const char *value) {
+  if (ignorelist == NULL) {
+    ignorelist = ignorelist_create(1);
+    if (ignorelist != NULL) {
```
Comment (kwiatrox) [inline src/buddyinfo.c]: if (ignorelist == NULL) {
  ERROR("buddyinfo plugin: ignorelist_create failed");
  return -ENOMEM; // be consistent  and return negative numbers on error
}

Please test your commits before submitting. In earlier comment I refered to (ignorelist != NULL) as a positive case.

File: src/buddyinfo.c [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+      WARNING("line %s doesn't contain %d orders, skipping...", buffer,
+              MAX_ORDER);
+      continue;
+    }
+
+    zone = fields[1];
+    for (int i = 1; i <= MAX_ORDER; i++) {
+      ssnprintf(pagesize_kb, sizeof(pagesize_kb), "%dKB",
+                NUM_OF_KB(pagesize, i - 1));
+      buddyinfo_submit(zone, pagesize_kb, (int)atoll(fields[i + 1]));
```
Comment (kwiatrox) [inline src/buddyinfo.c]: If biggest expected output fits the int, why not use `atoi`?

**Comment by kwiatrox** [comment]
I think it is not much work to add node info to metrics. You can use a format like: 'Node0/DMA32', 'Node0/Normal', 'Node1/Normal' etc. for plugin-instance. I do not insist on filtering option.

File: src/buddyinfo.c [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+static int config_keys_num = STATIC_ARRAY_SIZE(config_keys);
+
+static ignorelist_t *ignorelist;
+
+static int buddyinfo_config(const char *key, const char *value) {
+  if (ignorelist == NULL) {
+    ignorelist = ignorelist_create(1);
+    if (ignorelist == NULL) {
+      ERROR("buddyinfo plugin: ignorelist_create failed");
+      sfree(ignorelist);
```
Comment (kwiatrox) [inline src/buddyinfo.c]: sfree has no effect and is not required here

**Pull request merged**